src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/FIPSTokenLoginHandler.java

```diff
@@ -0,0 +1,149 @@
+        if (!(callbacks[0] instanceof PasswordCallback)) {
+            throw new UnsupportedCallbackException(callbacks[0]);
+        }
+        PasswordCallback pc = (PasswordCallback)callbacks[0];
+        pc.setPassword(getFipsNssdbPin());
+    }
+
+    private static char[] getFipsNssdbPin() throws ProviderException {
+        if (debug != null) {
+            debug.println("FIPS: Reading NSS DB PIN for token...");
+        }
+        String pinProp = SecurityProperties
+                .privilegedGetOverridable(FIPS_NSSDB_PIN_PROP);
+        if (pinProp != null && !pinProp.isEmpty()) {
+            String[] pinPropParts = pinProp.split(":", 2);
+            if (pinPropParts.length < 2) {
+                throw new ProviderException("Invalid " + FIPS_NSSDB_PIN_PROP +
+                        " property value.");
+            }
+            String prefix = pinPropParts[0].toUpperCase();
+            String value = pinPropParts[1];
+            String pin = null;
+            if (prefix.equals("ENV")) {
+                if (debug != null) {
+                    debug.println("FIPS: PIN value from the '" + value +
+                            "' environment variable.");
+                }
+                pin = System.getenv(value);
+            } else if (prefix.equals("FILE")) {
+                if (debug != null) {
+                    debug.println("FIPS: PIN value from the '" + value +
+                            "' file.");
+                }
+                pin = getPinFromFile(Paths.get(value));
+            } else if (prefix.equals("PIN")) {
+                if (debug != null) {
+                    debug.println("FIPS: PIN value from the " +
+                            FIPS_NSSDB_PIN_PROP + " property.");
+                }
+                pin = value;
+            } else {
+                throw new ProviderException("Unsupported prefix for " +
+                        FIPS_NSSDB_PIN_PROP + ".");
+            }
+            if (pin != null && !pin.isEmpty()) {
+                if (debug != null) {
+                    debug.println("FIPS: non-empty PIN.");
+                }
+                /*
+                 * C_Login in libj2pkcs11 receives the PIN in a char[] and
+                 * discards the upper byte of each char, before passing
+                 * the value to the NSS Software Token. However, the
+                 * NSS Software Token accepts any UTF-8 PIN value. Thus,
+                 * expand the PIN here to account for later truncation.
+                 */
+                byte[] pinUtf8 = pin.getBytes(StandardCharsets.UTF_8);
+                char[] pinChar = new char[pinUtf8.length];
+                for (int i = 0; i < pinChar.length; i++) {
+                    pinChar[i] = (char)(pinUtf8[i] & 0xFF);
+                }
+                return pinChar;
+            }
+        }
+        if (debug != null) {
+            debug.println("FIPS: empty PIN.");
+        }
+        return new char[] {};
```

Comment:
The non-FIPS `provider.login(null, null)` statement here [1] cannot be directly compared to the return statement here [2] because the former implies not calling PKCS#11 `C_Login` in the token [3] (which is fine if the NSS DB does not require a login), while the latter implies calling it with a PIN of len == 0 [4] (which must be done if the NSS DB requires a login). The effect of returning null or empty-array here [2] is whether we call `C_Login` with a NULL pPin parameter here [5] or not, but in both cases `ulPinLen` is 0 and that's what matters for our default FIPS scenario. Given that the effect is the same but we can save some performance penalty by not creating an empty array, I will make the change.

Let me take this opportunity to give some context and explain why we want to always do a login in FIPS mode (which is calling `provider::login` with a non-null 2nd parameter) [6].

If an NSS DB is not configured to require a login, the NSS Software Token cannot operate in FIPS-140 level 2 [7]. Please keep in mind that OpenJDK and the NSS Software Token must operate in FIPS-140 level 2 when RHEL is in FIPS mode. Thus, not making a login is not right.

There are 2 ways in which an NSS DB can require a login: not having a PIN set (think of this as an unitialized PIN) [8] or having a PIN set to something different than empty string [9]. Thus, having an empty string PIN set in the NSS DB does not require a login. When the NSS DB requires a login, the internal `CKF_LOGIN_REQUIRED` flag is set.

RHEL's deafult NSS DB (`/etc/pki/nssdb`) does not have a PIN set. Thus, the `CKF_LOGIN_REQUIRED` flag is set and a login is required. To login in this DB we must use an empty string PIN. In other words, `/etc/pki/nssdb` requires an empty string login, which is calling the PKCS#11 `C_Login` API with `usPinLen == 0` [10], and the underlying `NSC_Login` with `ulPinLen == 0` [5].

--
[1] - https://github.com/rh-openjdk/jdk/pull/17/files#diff-fb2c3293fab6ce24717b91fcf63fb9b33989bce9518910faf46e086493c753a3R247
[2] - https://github.com/rh-openjdk/jdk/pull/17/files#diff-dfa0dff6275fe003a6a2b68903fe3b3d9b0fac72d988f50a3554cd533c9efbd9R122
[3] - https://github.com/rh-openjdk/jdk/blob/f096804b619833f126a48c01da590ecd37d9cd67/src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java#L1668
[4] - https://github.com/rh-openjdk/jdk/blob/f096804b619833f126a48c01da590ecd37d9cd67/src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java#L1716
[5] - https://github.com/nss-dev/nss/blob/NSS_3_67_RTM/lib/softoken/pkcs11.c#L4289
[6] - https://github.com/rh-openjdk/jdk/pull/17/files#diff-fb2c3293fab6ce24717b91fcf63fb9b33989bce9518910faf46e086493c753a3R244
[7] - https://github.com/nss-dev/nss/blob/NSS_3_67_RTM/lib/softoken/fipstokn.c#L638
[8] - https://github.com/nss-dev/nss/blob/NSS_3_67_RTM/lib/softoken/pkcs11.c#L3720
[9] - https://github.com/nss-dev/nss/blob/NSS_3_67_RTM/lib/softoken/pkcs11.c#L3724
[10] - https://github.com/nss-dev/nss/blob/NSS_3_67_RTM/lib/softoken/fipstokn.c#L876